HW2/P5/physics.pyx

```diff
@@ -69,18 +78,37 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
 
+    grid_size = <unsigned int> (1.0 / grid_spacing + 1)
+    gx = <unsigned int> (XY[i,0] / grid_spacing)
+    gy = <unsigned int> (XY[i,1] / grid_spacing)
+
+    # iterate through a 5 x 5 square on the grid
+    for j in range(gx - 2, gx + 3):
+        for k in range(gy - 2, gy + 3):
+            # ensure j,k are in the grid, and only collide with larger indices (avoid double collisions)
+            if ((0 <= j < grid_size) and (0 <= k < grid_size) and (i < Grid[j,k] < count)):
```

Comment:
This is a very unique and clever way of checking if j and k are in bounds. I like this implementation of checking if both j and k are greater than 0 and less than the grid size. Very nicely done!